### ISSUE TITLE
"RecursionError: maximum recursion depth exceeded" with pandas 1.1.5

### ISSUE THREAD

**Issue by freud14** [PyCQA/pylint#3977]
As discussed in issue #3939.

### Steps to reproduce

It seems that with the new version 1.1.5 of pandas, we get a recursion error when using pandas in a similar way to issue #3939. I use the same script as @arg-123 in issue #3939:
```python
import pandas as pd

def func(df1, df2):
    return pd.merge(df1, df2)  # RecursionError with pylint!

df = pd.DataFrame([[0]])
func(df, df)
```
Here is the bash script I use to reproduce the bug:
```bash
python3.8 -m venv env
source env/bin/activate
pip install pandas pylint # pandas==1.1.5 pylint==2.6
pip install wheel
pip install git+https://github.com/PyCQA/astroid.git
pylint test.py
```
where `test.py` is the Python script above.

With pandas 1.1.4, the bug is not present.

### Current behavior

Recursion error, see https://pastebin.com/EXJJSiqt

### Expected behavior

No recursion error

### pylint --version output

```
pylint 2.6.0
astroid 2.5.0
Python 3.8.5 (default, Jul 28 2020, 12:59:40) 
[GCC 9.3.0]
```

**Comment by bnsh** [comment]
Interesting that you say that with 1.1.4 the bug is not present. I tried with the python:3.8 docker image:

and I get the same recursion error. Perhaps I'll try to find which pandas version will finally allow me to get it working.

```
FROM python:3.8

MAINTAINER Binesh Bannerjee <binesh_binesh@hotmail.com>

RUN /usr/local/bin/python3.8 -m pip install -U pylint pandas==1.1.4

RUN echo "import pandas as pd" > /tmp/demo.py && \
    echo "pd.merge_asof()" >> /tmp/demo.py

RUN /usr/local/bin/python3.8 -m pip freeze

CMD ["/usr/local/bin/python3.8", "-m", "pylint", "-r", "n", "/tmp/demo.py"]

# docker build -t binesh/pylintbug .
# docker run -it --rm binesh/pylintbug
#
# demo.py simply contains
# import pandas as pd
# pd.merge_asof()
#
# pip freeze output:
# astroid==2.4.2
# isort==5.6.4
# lazy-object-proxy==1.4.3
# mccabe==0.6.1
# numpy==1.19.4
# pandas==1.1.4
# pylint==2.6.0
# python-dateutil==2.8.1
# pytz==2020.4
# six==1.15.0
# toml==0.10.2
# wrapt==1.12.1


```

**Comment by bnsh** [comment]
pandas==1.0.5 finally runs to completion, with these (correct) python warnings:

```
/tmp/demo.py:1:0: C0114: Missing module docstring (missing-module-docstring)
/tmp/demo.py:2:0: E1120: No value for argument 'left' in function call (no-value-for-parameter)
/tmp/demo.py:2:0: E1120: No value for argument 'right' in function call (no-value-for-parameter)
```

**Comment by freud14** [comment]
@bnsh it's because you are not at the latest version of astroid, i.e. master branch on GitHub.

**Comment by hippo91** [comment]
Thanks to @dbaty, :+1:  i made three kind of tests (all with `python 3.8.6`).
First one with `pandas 1.1.4`: the problem is no more present if i use `astroid`'s master branch;
Second one with `pandas 1.1.5`: the problem is present even with `astroid`'s master branch;
Third one with last pandas version (1.2.0): the problem is not present even with current version of `astroid`.
Thus i suggest moving from `pandas 1.1.4` to `pandas 1.2.0`.
I close this issue, but will reopen it if needed.

**Comment by freud14** [comment]
@hippo91 Instead of closing this issue, wouldn't it be better to find the underlying issue in astroid? It seems to me that there are a lot of bug reports with recursion error in astroid and a new version of pandas does not solve the bug that will reappear in a future version or in other libraries.
